bundles/org.openhab.binding.nobohub/src/main/java/org/openhab/binding/nobohub/internal/connection/HubCommunicationThread.java

```diff
@@ -0,0 +1,162 @@
+
+import java.time.Duration;
+import java.time.Instant;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.nobohub.internal.NoboHubBindingConstants;
+import org.openhab.binding.nobohub.internal.NoboHubBridgeHandler;
+import org.openhab.binding.nobohub.internal.model.NoboCommunicationException;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thread that reads from the Nobø Hub and sends HANDSHAKEs to keep the connection open.
+ *
+ * @author Jørgen Austvik - Initial contribution
+ */
+@NonNullByDefault
+public class HubCommunicationThread extends Thread {
+
+    private enum HubCommunicationThreadState {
+        STARTING(ThingStatus.ONLINE, ThingStatusDetail.UNKNOWN.NONE, "@text/message.bridge.status.connecting"),
+        CONNECTED(ThingStatus.ONLINE, ThingStatusDetail.NONE, ""),
+        DISCONNECTED(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "@text/message.bridge.status.failed"),
+        STOPPED(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, "");
+
+        private final ThingStatus status;
+        private final ThingStatusDetail statusDetail;
+        private final String errorMessage;
+
+        HubCommunicationThreadState(ThingStatus status, ThingStatusDetail statusDetail, String errorMessage) {
+            this.status = status;
+            this.statusDetail = statusDetail;
+            this.errorMessage = errorMessage;
+        }
+
+        public ThingStatus getThingStatus() {
+            return status;
+        }
+
+        public ThingStatusDetail getThingStatusDetail() {
+            return statusDetail;
+        }
+
+        public String getErrorMessage() {
+            return errorMessage;
+        }
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(HubCommunicationThread.class);
+
+    private final HubConnection hubConnection;
+    private final NoboHubBridgeHandler hubHandler;
+    private final Duration timeout;
+    private Instant lastTimeFullScan;
+
+    private volatile boolean stopped = false;
+    private HubCommunicationThreadState currentState = HubCommunicationThreadState.STARTING;
+
+    public HubCommunicationThread(HubConnection hubConnection, NoboHubBridgeHandler hubHandler, Duration timeout) {
+        this.hubConnection = hubConnection;
+        this.hubHandler = hubHandler;
+        this.timeout = timeout;
+        this.lastTimeFullScan = Instant.now();
+    }
+
+    public void stopNow() {
+        stopped = true;
+    }
+
+    @Override
+    public void run() {
+        while (!stopped) {
+            switch (currentState) {
+                case STARTING:
+                    try {
+                        hubConnection.refreshAll();
+                        lastTimeFullScan = Instant.now();
+                        setNextState(HubCommunicationThreadState.CONNECTED);
```

Comment:
Isn't this fighting with `NoboHubBridgeHandler.initialize()`?

```java
                HubCommunicationThread ht = new HubCommunicationThread(conn, this, timeout);
                ht.start();
                hubThread = ht;

                if (ht.getConnection().isConnected()) {
                    logger.debug("Communication thread to {} is up and running, we are online", hostName);
                    updateProperty(Thing.PROPERTY_SERIAL_NUMBER, serialNumber);
                    updateStatus(ThingStatus.ONLINE);
                } else {
                    logger.debug("HubCommunicationThread is not connected anymore, setting to OFFLINE");
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                            "@text/message.bridge.connection.failed");
                }
```

If I read it correctly, a new thread will be started, and this thread will by itself update status. In the meantime the handler will proceed and check if connected, and also set status accordingly. Most likely it will not be connected yet, although this depends on timing? Perhaps you could paste the logged status transitions during startup?

Reply:
Here is an example where I added a logging line for the update of the status:

    2022-08-21 21:01:46.297 [DEBUG] [obohub.internal.NoboHubBridgeHandler] - Done connecting to 10.0.0.203 (102000060208)
    2022-08-21 21:01:46.297 [DEBUG] [obohub.internal.NoboHubBridgeHandler] - Starting communication thread to 10.0.0.203
    2022-08-21 21:01:46.297 [DEBUG] [obohub.internal.NoboHubBridgeHandler] - Communication thread to 10.0.0.203 is up and running, we are online
    2022-08-21 21:01:46.506 [DEBUG] [obohub.internal.NoboHubBridgeHandler] - Updating Hub: My Eco Hub
    2022-08-21 21:01:46.506 [DEBUG] [obohub.internal.NoboHubBridgeHandler] - Updating Hub with ActiveOverrideId 101 with Name NORMAL
    ...
    2022-08-21 21:01:46.512 [DEBUG] [obohub.internal.NoboHubBridgeHandler] - Updating Nobø Hub Status to 'ONLINE' detail 'NONE' and description '' 
    2022-08-21 21:01:54.466 [DEBUG] [al.connection.HubCommunicationThread] - HubCommunicationThread is stopped, disconnecting from Hub

The actual connection is done a few lines above the exampled code, in line:

`conn.connect();`

Then the **conn** object is passed on to the HubCommunicationThread, which has the ability to reconnect if needed. This might still be a timing issue, but the code in HubCommunicationThread has the ability to do a reconnect if the initial connection should fail. As discussed in an earlier comment, I also think the approach used by some of the other bindings do look a bit safer regarding timing issues, so I think we might look into refactoring this for later. For now I don't think this is a big issue, but it might explain the few connection issues I have seen once or twice a month.

Reply:
The framework also logs all `updateStatus()`, and here I suspect you might see some additional updates like:
- UNKNOWN (NoboHubBridgeHandler)
- OFFLINE (NoboHubBridgeHandler) - because not connected yet and thread is async. Serial number property skipped.
- ONLINE (HubCommunicationThread)

Or:
- UNKNOWN (NoboHubBridgeHandler)
- ONLINE (HubCommunicationThread)
- ONLINE (NoboHubBridgeHandler) - and now serial number property set.

It looks like a timing issue/bug, so if you are working on a fix and prefer to merge this PR before fixing it, I'm okay with that since it seems to be a minor issue.

Reply:
Redeploy of jar:

    2022-08-21 22:16:06.004 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from ONLINE to UNINITIALIZED
    2022-08-21 22:16:06.008 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)
    2022-08-21 22:16:06.153 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING
    2022-08-21 22:16:06.155 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from INITIALIZING to UNKNOWN
    2022-08-21 22:16:06.634 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNKNOWN to ONLINE
    2022-08-21 22:16:06.831 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from ONLINE to ONLINE

Disable/Enable in UI:

    2022-08-21 22:15:33.076 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from ONLINE to UNINITIALIZED
    2022-08-21 22:15:33.080 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNINITIALIZED to UNINITIALIZED (DISABLED)
    2022-08-21 22:15:36.216 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNINITIALIZED (DISABLED) to INITIALIZING
    2022-08-21 22:15:36.218 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from INITIALIZING to UNKNOWN
    2022-08-21 22:15:36.681 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNKNOWN to ONLINE
    2022-08-21 22:15:36.895 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from ONLINE to ONLINE

Reply:
That seems consistent with my theory. Now you could try to add a small delay in `HubCommunicationThread` before connecting (to simulate slower network) to see if the first scenario would happen and how it would play out.

Reply:
A two second delay gives this log:

    2022-08-22 19:58:57.849 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'nobohub:nobohub:home' updated: UNINITIALIZED
    2022-08-22 19:58:57.849 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from ONLINE to UNINITIALIZED
    2022-08-22 19:58:57.853 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'nobohub:nobohub:home' updated: UNINITIALIZED (HANDLER_MISSING_ERROR)
    2022-08-22 19:58:57.853 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)
    2022-08-22 19:58:58.011 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'nobohub:nobohub:home' updated: INITIALIZING
    2022-08-22 19:58:58.011 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING
    2022-08-22 19:58:58.013 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'nobohub:nobohub:home' updated: UNKNOWN
    2022-08-22 19:58:58.013 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from INITIALIZING to UNKNOWN
    2022-08-22 19:58:58.348 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'nobohub:nobohub:home' updated: ONLINE
    2022-08-22 19:58:58.349 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from UNKNOWN to ONLINE
    2022-08-22 19:59:00.570 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'nobohub:nobohub:home' updated: ONLINE
    2022-08-22 19:59:00.571 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'nobohub:nobohub:home' changed from ONLINE to ONLINE

The delay adds the time difference between 19:58:58.349-> 19:59:00.570. As the actual connection is done in **conn.connect();**, the HubCommunicationThread is not created or started before after conn.connect with it`s org.openhab.binding.nobohub.internal.connection.HubConnection#connectSocket has successfully made a connection to the hub. When and if the connection is a success, the conn object is passed on to HubCommunicationThread, which has the job to retry and keep the connection going. The two second delay is then added before we get to 

                if (ht.getConnection().isConnected()) {

which explains why the updateStatus is delayed by two seconds. 

I could be wrong, but I don`t think there will be timing issue here unless HubCommunicationThread is delayed in any way. A small improvement could be to check **isConnected()** before actually creating HubCommunicationThread, this might also improve the readability of how this mechanism works.

Reply:
OK, I was afraid that `ht.getConnection().isConnected()` would return false when called right after thread is started while not yet connected. I totally missed that the sync `conn.connect();` is actually called within the bridge handler, not within  `HubCommunicationThread`. Sorry for the confusion! However, `disconnect()` is called within `HubCommunicationThread`, so seems like responsibilities are a bit divided. Nevertheless, I was chasing a ghost. Only slight problem is the double setting of status **ONLINE**. I'm resolving this comment, and since there are no more open comments, I will merge within an hour or so unless you object. :-)

Reply:
Merge it!